### PR TITLE
support type synonyms in scala

### PR DESCRIPTION
**Pull request by nickchapman-da**
Support DAML-LF type synonyms in scala code base.
- AST
- decode
- type check
- extended DAML-LF parser, to allow:
- testing of synonym expansion during type checking

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by cocreature** [comment]
There are a bunch of typechecking tests in https://github.com/digital-asset/daml/blob/master/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala. We should add some for type synonyms there as well.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ValueTranslator.scala [hunk 1]
```diff
@@ -51,6 +51,7 @@ private[engine] class ValueTranslator(compiledPackages: CompiledPackages) {
 
       def go(typ: Type): Type =
         typ match {
+          case TSynApp(_, _) => throw new RuntimeException("TODO #3616,ValueTranslator")
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ValueTranslator.scala]: `ValueTranslator` translates incoming Ledger API values to speedy expressions so it only has to handle serializable types. Therefore, you can just error out here using `fail` like we do for `TForall` and `TStruct`.

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/AstRewriter.scala [hunk 2]
```diff
@@ -52,6 +52,7 @@ private[digitalasset] class AstRewriter(
     if (typeRule.isDefinedAt(x)) typeRule(x)
     else
       x match {
+        case TSynApp(_, _) => throw new RuntimeException("TODO #3616,AstRewriter,TSynApp")
```
Comment (cocreature) [inline daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/AstRewriter.scala]: The `AstRewriter` is only used in `EncodeV1` so those two should be fixed together and as mentioned above, we can leave that for a separate PR.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Collision.scala [hunk 3]
```diff
@@ -68,6 +68,10 @@ private[validation] object Collision {
         // ignore values
         // List(NValDef(module, defName, vDef))
         List.empty
+
+      case _: Ast.DTypeSyn =>
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Collision.scala]: Happy to leave that for a separate PR.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala [hunk 4]
```diff
@@ -37,7 +37,8 @@ private[validation] object Serializability {
 
     def checkType(): Unit = checkType(typeToSerialize)
 
-    def checkType(typ0: Type): Unit = typ0 match {
+    def checkType(typ0: Type): Unit = typ0 match { //TODO: fix the exhaustiveness checking
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: I don’t understand this comment, aren’t you already handling the case that you added?

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala [hunk 5]
```diff
@@ -37,7 +37,8 @@ private[validation] object Serializability {
...
+      case TSynApp(syn,_) => unserializable(URDataType(syn))
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: In the Haskell typechecker you throw `URTypeSyn` instead of `URDataType`. I think we should be consistent here.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 6]
```diff
@@ -439,6 +444,35 @@ private[validation] object Typing {
         KStar
     }
 
+    private def expandTypeSynonyms(typ0: Type): Type = typ0 match {
+      case TSynApp(syn, args) =>
+        expandSynApp(syn, args)
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: In the Haskell typechecker, you recurse on the result of the expansion to expand synonyms that might be in there. Is this not necessary here? If so why?

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 7]
```diff
@@ -145,6 +145,26 @@ private[archive] class DecodeV1(minor: LV.Minor) extends Decode.OfPackage[PLF.Pa
       val defs = mutable.ArrayBuffer[(DottedName, Definition)]()
       val templates = mutable.ArrayBuffer[(DottedName, Template)]()
 
+      if (!onlySerializableDataDefs) {
+        // collect type synonyms
+        lfModule.getSynonymsList.asScala
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala]:
```suggestion
        assertSince(LV.Feature.typeSynonyms) 
        lfModule.getSynonymsList.asScala
```

assuming we add in `daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/LanguageVersion.scala` the following line:

``
    val typeSynonyms = v1_dev
``

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/PackageLookup.scala [hunk 8]
```diff
@@ -17,6 +17,9 @@ object PackageLookup {
       dataTyp <- defn match {
         case _: DValue =>
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/PackageLookup.scala]: Could you put the error cases at the end of the pattern matching ?

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ValueTranslator.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ValueTranslator.scala]: @cocreature is right. Just handle this case as `TForall` and `TStruct`.

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala [hunk 9]
```diff
@@ -272,6 +272,7 @@ object Repl {
 
   def prettyDefinitionType(defn: Definition, pkgId: PackageId, modId: ModuleName): String =
     defn match {
+      case DTypeSyn(_, _) => "<type synonym>" // FIXME(JM): pp this
```
Comment (remyhaemmerle-da) [inline daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala]: I am fine not pretty print the type synonyms, but you probably should not the initals of @dajmaki in your comment.

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala [hunk 10]
```diff
@@ -295,6 +296,7 @@ object Repl {
       if (needParens) s"($s)" else s
 
     def prettyType(t0: Type, prec: Int = precTForall): String = t0 match {
+      case TSynApp(_, _) => "<TODO: #3616,typeSynApp>"
```
Comment (remyhaemmerle-da) [inline daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala]: This pretty printer is not very important. 
It should be fine to inline `TSynApp(tyCon, args)` as `TApp(TyCon(tycon), args)`

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala [hunk 11]
```diff
@@ -15,6 +15,10 @@ private[validation] object TypeSubst {
 
   private def go(fv0: Set[TypeVarName], subst0: Map[TypeVarName, Type], typ0: Type): Type =
     typ0 match {
+      case TSynApp(syn, args) =>
+        TSynApp(syn, args.map { ty =>
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala]:
```suggestion
        TSynApp(syn, args.map(go(fv0, subst0, _))
```

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/ExprTraversable.scala [hunk 12]
```diff
@@ -118,6 +118,7 @@ private[validation] object ExprTraversable {
 
   private[traversable] def foreach[U](x: Definition, f: Expr => U): Unit =
     x match {
+      case DTypeSyn(params @ _, typ @ _) =>
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/ExprTraversable.scala]: :+1:

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala [hunk 13]
```diff
@@ -15,6 +15,8 @@ private[validation] object TypeTraversable {
 
   private[validation] def foreach[U](typ: Type, f: Type => U): Unit =
     typ match {
+      case TSynApp(_, args) =>
+        args.iterator.foreach(f)
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]:
```suggestion
        f(TyCon(tyCon))
        args.iterator.foreach(f)
```

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala [hunk 14]
```diff
@@ -15,6 +15,8 @@ private[validation] object TypeTraversable {
 
   private[validation] def foreach[U](typ: Type, f: Type => U): Unit =
     typ match {
+      case TSynApp(_, args) =>
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]:
```suggestion
      case TSynApp(tyCon, args) =>
```

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala [hunk 15]
```diff
@@ -15,6 +15,8 @@ private[validation] object TypeTraversable {
 
   private[validation] def foreach[U](typ: Type, f: Type => U): Unit =
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: This 2 changes below, follow the current philosophy of the TypeTraversable iterator utility.
Without those changes, `Recursion` check will broke.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 16]
```diff
@@ -145,6 +145,26 @@ private[archive] class DecodeV1(minor: LV.Minor) extends Decode.OfPackage[PLF.Pa
...
+              defn.getNameDname,
+              PLF.DefTypeSyn.NameCase.NAME_INTERNED_DNAME,
+              defn.getNameInternedDname,
+              "DefTypeSyn.name.name"
+            )
+            currentDefinitionRef =
+              Some(DefinitionRef(packageId, QualifiedName(moduleName, defName)))
+            val d = decodeDefTypeSyn(defn)
+            defs += (defName -> d)
+          //throw ParseError("type synonym def: " + defName.toString()) // TODO #3616 - remove
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala]: line just needs to be deleted. done

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1.scala [hunk 17]
```diff
@@ -74,6 +74,10 @@ private[digitalasset] class EncodeV1(val minor: LV.Minor) {
             }
           case value @ DValue(_, _, _, _) =>
             builder.addValues(name -> value)
+
+          case DTypeSyn(params @ _, typ @ _) =>
+            throw new RuntimeException("TODO #3616, EncodeV1, DTypeSyn")
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1.scala]: ok

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ValueTranslator.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ValueTranslator.scala]: ok, done

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/AstRewriter.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/AstRewriter.scala]: ok

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Collision.scala (hunk 3 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Collision.scala]: ok

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala (hunk 4 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: This comments is noting that the scala exhaustivity checking is not being done on this match expression. Which is sad. And it would be nice to get it fixed.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala (hunk 5 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: Thanks, nice catch. I wrote the `URTypeSyn` case class, but forgot to use it.

Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: fixed now

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 6 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: It is necessary here also. Thanks, nice catch. Fixed now.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala (hunk 7 above)
Comment (nickchapman-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala]: ok, I'll have a look at what needs doing here...

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1.scala [hunk 18]
```diff
@@ -205,7 +209,9 @@ private[digitalasset] class EncodeV1(val minor: LV.Minor) {
           case _ => typ0 -> ImmArray.empty
         }
       val builder = PLF.Type.newBuilder()
-      typ match {
+      typ match { //TODO: fix the exhaustiveness checking
+        case TSynApp(_, _) =>
+          throw new RuntimeException("TODO #3616,encodeTypeBuilder")
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1.scala]: ok, done

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/PackageLookup.scala (hunk 8 above)
Comment (nickchapman-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/PackageLookup.scala]: ok (but this was actually the way the code was before my changes!)
done

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala (hunk 9 above)
Comment (nickchapman-da) [reply daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala]: removed the JP

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala (hunk 10 above)
Comment (nickchapman-da) [reply daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala]: I filled in the PP code for this case

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala (hunk 11 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala]: ok, done

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/ExprTraversable.scala (hunk 12 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/ExprTraversable.scala]: Thank you. No code is the best code :)

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala (hunk 13 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: Are you sure? This looks wrong to me.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala (hunk 14 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: As above.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala (hunk 15 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: As above

Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: What will break? I don't understand

Comment (cocreature) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: @remyhaemmerle-da is referring to `modRefsInType` in `com.digitalasset.daml.lf.validation.Recursion` which currently checks for `TTyCon` and therefore will miss module references in a type synonym name. That said, I’m not sure I agree that the suggestions make sense here. Pretending that a type synonym is a type con is really confusing. I think it makes much more sense to extend `modRefsInType` to handle type synonyms.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala (hunk 4 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: Some more details: The match expression makes use of `unapply` methods defined for `ParametricType[12]` in 
daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Util.scala

Sadly, unapply methods cause the scala match-exhaustivity checker to get disabled (without any warning).

Comment (cocreature) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: scala is great …

Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: Maybe we should remove the TODO, but I think it worth keeping the comment that there is no exhaustivity check for these two match expressions.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala (hunk 15 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: Right. I see that now. 
And I agree a fix to `modRefsInType` would be clearer.

Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: Ok for the fix of  `modRefsInType`

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala (hunk 4 above)
Comment (cocreature) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: Happy to keep the comment. I would suggest to turn it into a non-TODO comment and make it a bit more explicit that this is about the scala exhaustiveness checker not checking this match rather and not this match not being exhaustive (at least at this point :slightly_smiling_face:)

Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Serializability.scala]: I made a separate PR to fix this case of lost exhaustivity checking. #4139

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala (hunk 13 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: we fixed `modRefsInType` instead

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala (hunk 15 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala]: done

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala [hunk 19]
```diff
@@ -25,6 +25,8 @@ private[validation] object Recursion {
     val modRefsInType: Set[ModuleName] = {
 
       def modRefsInType(acc: Set[ModuleName], typ0: Type): Set[ModuleName] = typ0 match {
+        case TSynApp(typeSynName,_) if typeSynName.packageId == pkgId =>
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: Don’t you need to recurse on the arguments here?

Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: Wont that happen in the `TypeTraversable`?

Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: Ah, no it wont. Think you're right.

Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: Actually, take 3. I'm not sure.
Too much scala magic for my brain to process :(

Comment (cocreature) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: My understanding is that `TypeTraversable` is roughly the equivalent of something like Haskell’s `plate`, i.e., a traversal for the immediate children that only goes down one level. Then it’s up to the function that you’re using to traverse to do further recursion, that’s why the `otherwise`case here calls the traversable again. Since you don’t fall into the otherwise case, you need to recurse on the arguments either by inlining the recursion on the arguments or my calling the traversable here as well.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala [hunk 20]
```diff
@@ -25,6 +25,8 @@ private[validation] object Recursion {
...
+          acc + typeSynName.qualifiedName.module
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]:
```suggestion
          TypeTraversable(typ0).fodLeft(acc + typeSynName.qualifiedName.module)(modRefsInType)
```

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala (hunk 19 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: `TypeTraversable(typp: Type)` gave you an traversable over the *direct* subtree (of type `Type`) of `typ0`.
If you want to iterate over all the subtree of `typ` you have to recurse on each subtree. 
Here that is done by the default case (line 32/33).
Since `TSynApp(typeSynName,_)` has its own case and contains subtree you have to recurse.

Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: :+1: for @cocreature explanation.

Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: Right. Thank you. I see now it is not doing deep recursion.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala (hunk 20 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: Thank you. I made the fix using `/:` to be consistent.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala (hunk 7 above)
Comment (nickchapman-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala]: This is done now. Thank you.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala (hunk 20 above)
Comment (SamirTalwar) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: [FYI, `/:` and `:\` are deprecated in favor of `foldLeft` and `foldRight` respectively.](https://www.scala-lang.org/api/2.12.10/scala/collection/TraversableOnce.html)

Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala]: All those `/:` are mine :( 
I will spend some time soon to remove all `/:`. 
You can use here whatever you prefer.

**Comment by nickchapman-da** [comment]
Yup. Testing is next.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 21]
```diff
@@ -265,6 +265,7 @@ private[validation] object Typing {
         }
       case (dfnName, dfn: DValue) =>
         Env(mod.languageVersion, world, ContextDefValue(pkgId, mod.name, dfnName)).checkDValue(dfn)
+      case (_, _: DTypeSyn) => // TODO #3616: check type synonyms
```
Comment (nickchapman-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: And completing this missing piece!

Comment (cocreature) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: Happy to do this in a separate PR. We should still be able to test expansion of type synonyms without checking that the definition is well formed

**Comment by nickchapman-da** [comment]
Tests for synonym expansion added to `TypingSpec.scala`

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/ModParser.scala [hunk 22]
```diff
@@ -56,6 +56,13 @@ private[parser] class ModParser[P](parameters: ParserParameters[P]) {
   private lazy val binder: Parser[(Name, Type)] =
     id ~ `:` ~ typ ^^ { case id ~ _ ~ typ => id -> typ }
 
+  private lazy val synDefinition: Parser[DataDef] =
+    Id("synonym") ~>! dottedName ~ rep(typeBinder) ~
```
Comment (cocreature) [inline daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/ModParser.scala]: Scala parser combinators are lovely …
